src/dataflow-types/src/client.rs

```diff
@@ -283,57 +311,287 @@ pub trait Client: Send {
+        assert_eq!(worker_txs.len(), worker_threads.len());
+        // assemble a list of process-local clients.
+        let mut locals = Vec::with_capacity(worker_txs.len());
+        for ((rx, tx), thread) in feedback_rxs.into_iter().zip(worker_txs).zip(worker_threads) {
+            locals.push(process_local::ProcessLocal::new(rx, tx, thread));
+        }
+        LocalClient {
+            client: partitioned::Partitioned::new(locals),
+        }
     }
+}
 
+#[async_trait::async_trait]
+impl Client for LocalClient {
     async fn send(&mut self, cmd: Command) {
-        log::trace!("Broadcasting dataflow command: {:?}", cmd);
-        let cmd_parts = cmd.partition_among(self.num_workers());
-        for (tx, cmd_part) in self.worker_txs.iter().zip(cmd_parts) {
-            tx.send(cmd_part)
-                .expect("worker command receiver should not drop first")
+        log::trace!("SEND dataflow command: {:?}", cmd);
+        self.client.send(cmd).await
+    }
+    async fn recv(&mut self) -> Option<Response> {
+        let response = self.client.recv().await;
+        log::trace!("RECV dataflow response: {:?}", response);
+        response
+    }
+}
+
+/// Clients whose implementation is partitioned across a set of subclients (e.g. timely workers).
+pub mod partitioned {
+
+    use std::collections::HashMap;
+
+    use expr::GlobalId;
+    use repr::Timestamp;
+
+    use super::Client;
+    use super::{ClientAsStream, Command, PeekResponse, Response};
+
+    /// A client whose implementation is sharded across a number of other clients.
+    ///
+    /// Such a client needs to broadcast (partitioned) commands to all of its clients,
+    /// and await responses from each of the client shards before it can respond.
+    pub struct Partitioned<C: Client> {
+        shards: Vec<C>,
+        state: PartitionedClientState,
+    }
+
+    impl<C: Client> Partitioned<C> {
+        /// Create a client partitioned across multiple client shards.
+        pub fn new(shards: Vec<C>) -> Self {
+            Self {
+                state: PartitionedClientState::new(shards.len()),
+                shards,
+            }
         }
-        for thread in &self.worker_threads {
-            thread.unpark()
+    }
+
+    #[async_trait::async_trait]
+    impl<C: Client> Client for Partitioned<C> {
+        async fn send(&mut self, cmd: Command) {
+            let cmd_parts = cmd.partition_among(self.shards.len());
+            for (shard, cmd_part) in self.shards.iter_mut().zip(cmd_parts) {
+                shard.send(cmd_part).await;
+            }
+        }
+
+        async fn recv(&mut self) -> Option<Response> {
+            use futures::StreamExt;
+            let mut stream = futures::stream::select_all(self.shards.iter_mut().enumerate().map(
+                |(index, client)| ClientAsStream { client }.map(move |response| (index, response)),
+            ));
+            while let Some((index, response)) = stream.next().await {
```

Comment:
Is it possible that instead of going to all this trouble to make streams, you just want `future::select_all`?

```diff
diff --git a/src/dataflow-types/src/client.rs b/src/dataflow-types/src/client.rs
index a5d0e2a73..e4f58ef5e 100644
--- a/src/dataflow-types/src/client.rs
+++ b/src/dataflow-types/src/client.rs
@@ -321,21 +321,6 @@ impl Client for Box<dyn Client> {
     }
 }
 
-/// A helper struct which allows us to interpret a `Client` as a `Stream` of `Response`.
-pub struct ClientAsStream<'a, C: Client + 'a> {
-    client: &'a mut C,
-}
-
-use std::pin::Pin;
-use std::task::{Context, Poll};
-impl<'a, C: Client + 'a> futures::stream::Stream for ClientAsStream<'a, C> {
-    type Item = Response;
-    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
-        use futures::Future;
-        Pin::new(&mut self.client.recv()).poll(cx)
-    }
-}
-
 /// A convenience type for compatibility.
 pub struct LocalClient {
     client: partitioned::Partitioned<process_local::ProcessLocal>,
@@ -378,11 +363,13 @@ pub mod partitioned {
 
     use std::collections::HashMap;
 
+    use futures::future;
+
     use expr::GlobalId;
     use repr::Timestamp;
 
     use super::Client;
-    use super::{ClientAsStream, Command, PeekResponse, Response};
+    use super::{Command, PeekResponse, Response};
 
     /// A client whose implementation is sharded across a number of other clients.
     ///
@@ -413,11 +400,9 @@ pub mod partitioned {
         }
 
         async fn recv(&mut self) -> Option<Response> {
-            use futures::StreamExt;
-            let mut stream = futures::stream::select_all(self.shards.iter_mut().enumerate().map(
-                |(index, client)| ClientAsStream { client }.map(move |response| (index, response)),
-            ));
-            while let Some((index, response)) = stream.next().await {
+            while let (Some(response), index, _) =
+                future::select_all(self.shards.iter_mut().map(|shard| shard.recv())).await
+            {
                 let message = self.state.absorb_response(index, response);
                 if let Some(message) = message {
                     return Some(message);
```

Maybe some control flow issues I haven't thought about?

Reply:
It's certainly not something that I had thought about. I'm mostly copy/pasting existing examples rather than understanding smart patterns. It seems like (maybe?) one difference is that if we expect to produce and absorb many responses without producing a response of our own, then doing `stream::select_all` once and draining it, rather than `future::select_all` each time around the loop, might be more efficient?

But short answer is "nope, didn't intentionally pick this".

Reply:
If I understand correctly, the `futures` approach does have the appealing property that it can see `None` responses, which is what (might) allow us to notice terminated replicas in the replicated setting. Maybe I'll try this out there, and we can see what it looks like, whether there are ergo issues, idk.